specification/flag-evaluation/flag-evaluation.md

```diff
@@ -0,0 +1,195 @@
+
+##### Requirement 1.13
+
+> The `evaluation details` wrapper **MUST** have a `flag key` field field of type string, representing the flag key.
+
+##### Requirement 1.14
+
+> The `evaluation details` wrapper **MUST** have an optional `variant` field of type string which represents a space-constrained identifier for a particular flag value.
+
+##### Requirement 1.15
+
+> The `evaluation details` wrapper **MUST** have a `reason` field of type string indicating the semantic reason for the returned flag value. Possible values vary by provider, but may include `"TARGETING_MATCH"`, `"SPLIT"`, `"DISABLED"`, `"DEFAULT"`, `"UNKNOWN"` and `"ERROR"`.
+
+##### Requirement 1.16
+
+> In cases of abnormal execution (network failure, unhandled error, etc) the `reason` field in the `evaluation details` **MUST** be set to `"ERROR"`.
+
+##### Requirement 1.17
+
+> The `evaluation details` wrapper **MUST** have an optional `error code` field of type string, which, in cases of abnormal execution, identifies an error occurred during flag evaluation, with possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.
+
+##### Requirement 1.18
+
+> If the `reason` in the `evaluation details` is set to `"ERROR"`, the sibling `error code` field **MUST** be set.
+
+##### Requirement 1.19
+
+> The `evaluation details` wrapper **MUST** have a `executed hooks` field, containing an object which itself contains 4 arrays, identified with keys (`before`, `after`, `error` and `finally`) indicating the hooks run for the respective life-cycle stages.
+
+```
+executedHooks: {
+  before: [],
+  after: [],
+  error: [],
+  finally: []
+}
+```
+
+##### Requirement 1.20
+
+> The value of the elements in the arrays contained in the `executed hooks` object **SHOULD** allow authors to uniquely identify the hooks executed during the flag evaluation lifecycle. This could be done with a mandatory string field defined on the `hook` interface, or by using the fully qualified class name in languages where such concepts exist.
+
+##### Requirement 1.21
+
+> The `evaluation options` object **MUST** have a `hooks` field which can contain hooks to be executed for the respective flag evaluation.
+
+see [hooks](./hooks.md)
+
+##### Requirement 1.22
+
+> No methods, functions, or operations on the client should ever throw exceptions, or otherwise abnormally end. Flag evaluation calls must always return the `default value` in the event of abnormal execution.
+
+##### Condition 1.23
+
+> The language type system differentiates between floating and integer type numeric values.
+>
+> ##### Conditional Requirement 1.23.1
+>
+> > The `client` **MUST** provide functions for the resolution of both floating point and integer numeric values.
+
+##### Condition 1.24
+
+> The language features async/await semantics.
+>
+> ##### Conditional Requirement 1.24.1
+>
+> > The `client` **SHOULD** provide asynchronous functions for flag evaluation.
```

Comment:
This is extremely specific. I would generalize it to something like "The client MUST provide a function for flag evaluation which does not block the thread of execution". I would also not gate this on async/await specifically because that is only one of many ways to achieve this goal.

Reply:
Largely agreed, but that would be changing a "SHOULD" to a "MUST". Is that what you recommend? In multithreaded environments like Java, blocking probably isn't that uncommon.

Reply:
Whether you want SHOULD or MUST, the important part is not the async/await syntax but the execution blocking